lib/OpenQA/WebAPI/Plugin/ObsRsync.pm

```diff
@@ -27,30 +29,101 @@ sub register {
+          ->to('Plugin::ObsRsync::Controller::Gru#index');
+        $plugin_r->post('/obs_rsync/#folder/runs')->name('plugin_obs_rsync_run')
+          ->to('Plugin::ObsRsync::Controller::Gru#run');
+
         $plugin_r->get('/obs_rsync/#folder/runs/#subfolder/download/#filename')->name('plugin_obs_rsync_download_file')
-          ->to('Plugin::ObsRsync::Controller#download_file');
-        $plugin_r->get('/obs_rsync/#folder/runs/#subfolder')->name('plugin_obs_rsync_logfiles')
-          ->to('Plugin::ObsRsync::Controller#logfiles');
-        $plugin_r->get('/obs_rsync/#folder/runs')->name('plugin_obs_rsync_logs')
-          ->to('Plugin::ObsRsync::Controller#logs');
+          ->to('Plugin::ObsRsync::Controller::Folders#download_file');
+        $plugin_r->get('/obs_rsync/#folder/runs/#subfolder')->name('plugin_obs_rsync_run')
+          ->to('Plugin::ObsRsync::Controller::Folders#run');
+        $plugin_r->get('/obs_rsync/#folder/runs')->name('plugin_obs_rsync_runs')
+          ->to('Plugin::ObsRsync::Controller::Folders#runs');
         $plugin_r->get('/obs_rsync/#folder')->name('plugin_obs_rsync_folder')
-          ->to('Plugin::ObsRsync::Controller#folder');
-        $plugin_r->get('/obs_rsync/')->name('plugin_obs_rsync_index')->to('Plugin::ObsRsync::Controller#index');
+          ->to('Plugin::ObsRsync::Controller::Folders#folder');
+        $plugin_r->get('/obs_rsync/')->name('plugin_obs_rsync_index')
+          ->to('Plugin::ObsRsync::Controller::Folders#index');
         $app->config->{plugin_links}{operator}{'OBS Sync'} = 'plugin_obs_rsync_index';
     }
 
     if (!$plugin_api_r) {
         $app->log->error('API routes not configured, plugin ObsRsync will not have API configured') unless $plugin_r;
     }
     else {
-        $plugin_api_r->put('/obs_rsync/#folder/runs')->name('plugin_obs_rsync_run')
-          ->to('Plugin::ObsRsync::Controller#run');
+        $plugin_api_r->put('/obs_rsync/#folder/runs')->name('plugin_obs_rsync_api_run')
+          ->to('Plugin::ObsRsync::Controller::Gru#run');
+    }
+
+    $app->minion->add_task(obs_rsync_run => sub { return OpenQA::WebAPI::Plugin::ObsRsync::Task::_run($app, @_) });
+}
+
+# try to determine whether project is dirty
+# undef means status is unknown
+sub _is_obs_project_status_dirty {
+    my ($url, $project) = @_;
+    return undef unless $url;
+
+    $url =~ s/%%PROJECT/$project/g;
+    my $ua = Mojo::UserAgent->new;
+
+    my $res = $ua->get($url)->result;
+    return undef unless $res->is_success;
+
+    return _parse_obs_response_dirty($res->body);
+}
+
+sub _parse_obs_response_dirty {
+    my ($body) = @_;
+
+    my $dirty;
+    return 1 if $body =~ /dirty/g;
+    while ($body =~ /^(.*repository="images".*)/gm) {
+        my $line = $1;
+        if ($line =~ /state="([a-z]+)"/) {
+            return 1 if $1 ne "published";
+            $dirty //= 0;
+        }
     }
+    return $dirty;
+}
+
+sub _check_and_render_error {
+    my $controller = $_[0];
```

Comment:
Method passes `@_` further, so cannot use `shift`

Reply:
That doesn't make any sense at all. You can just use `_check_error($c, @_)` or even better `_check_error($c, @args)`. Or since the function only uses the controller to call a helper, you could just pass along that value as `_check_error($c->obs_rsync->home, @args)`.

Reply:
OK, sorry if I step in here - but this really sounds borderline syntax taste. Can we get over this and merge or is there something else blocking?

Reply:
This is a minor issue, but i think it matters. Using `$_[0]` followed by `@_` on another line obfuscates a bit what `@_` still contains. If you assign `@_` to named variables instead there is no guessing anymore and it becomes obvious what is happening (and meant to happen) here.

Reply:
I got a bit sensitive to this kind of issue since i recently discovered stuff like `my ($foo, $bar) = (pop, pop);` in openQA.

Reply:
As @andrii-suse is in a training today and tomorrow and I want to see this deployed tomorrow, please fix the syntax and push to the branch